### PR TITLE
Unify Mesh::be_to_face

### PR DESCRIPTION
**Pull request by pazner**
Since #3379, `Mesh::be_to_face` was the mapping for boundary elements to "faces" (codim-1 entities) for 1D and 3D meshes, and `Mesh::be_to_edge` was the mapping for 2D meshes.

This PR removes `Mesh::be_to_edge` and simplifies some of the mesh logic.

~This PR adds 1D mesh support to `Mesh::GenerateBoundaryElements` and adds boundary elements to `ParSubMesh` in 1D.~

~(Previously, if one took a boundary submesh of a 2D mesh, no boundary elements would be created).~

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3811](https://github.com/mfem/mfem/pull/3811) | @pazner | @tzanio | @hughcars + @jandrej + @mlstowell | 8/12/23 | 11/4/23 | 11/9/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
Looks like #3379 (merged last week) also fixes this…

@mlstowell, it seems like in 1D, `be_to_face` is used for the boundary element to vertex mapping. Is there any reason to have a separate `be_to_edge` array in 2D? To me it looks like `be_to_edge` is only ever used in 2D, and `be_to_face` is used only in 3D (and now 1D). Perhaps `be_to_edge` can be removed (or deprecated for backwards compatibility if needed, though it is protected) and `be_to_face` can be used in all dimensions?

**Comment by jandrej** [comment]
Without looking at it yet, this should be compatible with #3461. No idea why the other PR is hovering for so long :)

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/mesh.hpp
Comment (termi-official) [inline mesh/mesh.hpp:1341]: Maybe it makes sense to deprecate this one, since it is equivalent to `GetBdrFace`?

Comment (pazner) [reply mesh/mesh.hpp:1341]: I'd be supportive of that. (The usage of the word "edge" in this function name is weird/confusing).

Comment (pazner) [reply mesh/mesh.hpp]: Done in 823eef1d9

**Comment by pazner** [comment]
I resolved conflicts and also made the deprecation suggested here: https://github.com/mfem/mfem/pull/3811#discussion_r1350838612

**Comment by v-dobrev** [comment]
Considering the presence of methods like `Mesh::GetBdrElement` and `Mesh::GetBdrElementFace`,
I think a more explicit name for `Mesh::GetBdrFace` will be better: e.g. `Mesh::GetBdrElementFaceIndex` or `Mesh::GetBdrElementFaceId`.

**Comment by pazner** [comment]
> Considering the presence of methods like `Mesh::GetBdrElement` and `Mesh::GetBdrElementFace`, I think a more explicit name for `Mesh::GetBdrFace` will be better: e.g. `Mesh::GetBdrElementFaceIndex` or `Mesh::GetBdrElementFaceId`.

Done in 9863bc5bd68ee9549be275c7e62777853f8e1f14.

File: mesh/mesh.cpp
Comment (hughcars) [inline mesh/mesh.cpp:6576]: Given this is now so simple, can possibly be inline in the header.

File: mesh/mesh.hpp
Comment (hughcars) [inline mesh/mesh.hpp:1363]: For the given boundary element no?

File: mesh/submesh/psubmesh.cpp
Comment (hughcars) [inline mesh/submesh/psubmesh.cpp:242]: If you'd like, you can bind this to a `const`:
```
const int num_codim_1 = [this]()
{
      if (Dim == 1) { return NumOfVertices; }
      else if (Dim == 2) { return NumOfEdges; }
      else if (Dim == 3) { return NumOfFaces; }
      else { MFEM_ABORT("Invalid dimension."); }
      return -1;
}();
```

**Review by hughcars — COMMENTED** [review]
Much prefer the new naming `GetBdrElementFaceIndex`!

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp]: Done in c794dae53b1d3723a4d1b0a39aa576d848a2a79e

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp]: Fixed in c794dae53b1d3723a4d1b0a39aa576d848a2a79e

File: mesh/submesh/psubmesh.cpp
Comment (pazner) [reply mesh/submesh/psubmesh.cpp:242]: Yeah, I sometimes use this idiom (["IIFE"](https://rigtorp.se/iife/)). It's good because of const, but it adds some "syntax noise" to the code (wouldn't it be great if C++ had "if-expressions"?). I'm fine either way.

Comment (hughcars) [reply mesh/submesh/psubmesh.cpp:242]: Yeah I guess I'm just very used to reading them now and like the const, I barely notice. Not worth holding the pr for, it's an aesthetic choice rather than functional.

**Review by hughcars — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: mesh/submesh/psubmesh.cpp
Comment (pazner) [reply mesh/submesh/psubmesh.cpp]: Done in 11aa88c5e

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
There are two issues with this PR (I believe) in tonight's autotest on tux:

- The following sample run segfaulted: `mpirun -np 4 ex14p -m ../data/inline-segment.mesh -rs 5 -no-vis`

- The GLVis build in `build-test-basic` complained about `GetBdrElementEdgeIndex` being deprecated. May be good to add a companion PR there.

**Comment by pazner** [comment]
> * The following sample run segfaulted: `mpirun -np 4 ex14p -m ../data/inline-segment.mesh -rs 5 -no-vis`

Should be fixed in 659432963. (Maybe the initialization logic could be refactored to make it easier to track these kinds of things).

> * The GLVis build in `build-test-basic` complained about `GetBdrElementEdgeIndex` being deprecated. May be good to add a companion PR there.

PR here: GLVis/glvis#266, but I think we should wait until this PR is merged first.

**Comment by tzanio** [comment]
Thanks!

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
From nightly testing: there are deprecation warnings in GLVis from this PR -- we need a GLVis  PR to fix them.

**Review by mlstowell — APPROVED** [review]

**Pull request merged**
